### PR TITLE
Add support for 32 bit Addresses in MCCAS debug info

### PR DESCRIPTION
**Pull request by rastogishubham**

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test llvm

File: llvm/lib/MCCAS/MCCASObjectV1.cpp [hunk 1]
```diff
@@ -843,6 +845,8 @@ getLineTableLengthInfoAndVersion(DWARFDataExtractor &LineTableDataReader,
     return std::move(Err);
 
   Prologue.PrologueLength = PrologueLength;
+  // DWARF 4 and below doesn't have the Address Size in the line table header,
+  // set the Offset and return.
```
Comment (felipepiovezan) [inline llvm/lib/MCCAS/MCCASObjectV1.cpp]: This comment feels a bit off given how this code path is _also_ used by DWARF 5. I understand you are trying to emphasize that this field is new in dwarf 5, but we already do this on line 824 above

**Comment by cachemeifyoucan** [comment]
High level comments:
* Make it clear this is for 32 bit debug info support in the commit message. Please make sure there is no limitation for 32 bit arch for TEXT/DATA sections.
* Make sure the correct `REQUIRES` is added for tests, tests require arm backends to be built
* When you test 32 bit arches, can you pick something more relevant than `armv4t`? Also it is always good to test `arm64_32` since it is such an outliner and might have edge cases.

File: llvm/lib/MCCAS/MCCASObjectV1.cpp [hunk 2]
```diff
@@ -1979,21 +1989,20 @@ struct DIEToCASConverter {
...
 
-Error InMemoryCASDWARFObject::partitionCUData(ArrayRef<char> DebugInfoData,
-                                              uint64_t AbbrevOffset,
-                                              DWARFContext *Ctx,
-                                              MCCASBuilder &Builder,
-                                              AbbrevSetWriter &AbbrevWriter,
-                                              uint16_t DwarfVersion) {
+Error InMemoryCASDWARFObject::partitionCUData(
+    ArrayRef<char> DebugInfoData, uint64_t AbbrevOffset, DWARFContext *Ctx,
+    MCCASBuilder &Builder, AbbrevSetWriter &AbbrevWriter, uint16_t DwarfVersion,
```
Comment (felipepiovezan) [inline llvm/lib/MCCAS/MCCASObjectV1.cpp]: If we have a MCCASBuilder, we have a MachOCASWriter, which has a Target; we can add a method `MCCASBuilder::getAddressSize` and avoid the needed for the redundant `AddressSize` parameter

File: llvm/lib/MCCAS/MCCASObjectV1.cpp [hunk 3]
```diff
@@ -3124,6 +3135,7 @@ struct DIEVisitor {
   DataExtractor DistinctExtractor;
   DataExtractor::Cursor DistinctCursor;
   StringRef DistinctData;
+  uint8_t AddressSize;
```
Comment (felipepiovezan) [inline llvm/lib/MCCAS/MCCASObjectV1.cpp]: `DataExtractor`s can have their AddressSize set, maybe we are (or should be?) setting it already?

**Review by felipepiovezan — COMMENTED** [review]
I think there are some places where we are redundantly passing the address size around,  but it's difficult to spot all of them, so I left some comments that may trigger "chain deletions" of diffs here!

File: llvm/include/llvm/MCCAS/MCCASObjectV1.h [hunk 4]
```diff
@@ -639,6 +639,8 @@ class MCCASReader {
 
   Triple::ArchType getArch() { return Target.getArch(); }
 
+  bool isArch32Bit() { return Target.isArch32Bit(); }
```
Comment (adrian-prantl) [inline llvm/include/llvm/MCCAS/MCCASObjectV1.h]: It seems like what you really want based on the uses is a getPointerSize() method instead?
```
Reader.isArch32Bit() ? 4 : 8)
```

File: llvm/test/CAS/watchos-cas.ll [hunk 5]
```diff
@@ -0,0 +1,49 @@
+; RUN: rm -rf %t && mkdir -p %t
+; RUN: llc --filetype=obj --mccas-verify --cas-backend --cas=%t/cas %s -o %t/watchos-cas.o 
+; RUN: llvm-dwarfdump %t/watchos-cas.o | FileCheck %s
+; CHECK: .debug_info contents:
+; CHECK-NEXT: 0x{{[0-9a-f]+}}: Compile Unit: length = 0x{{[0-9a-f]+}}, format = DWARF32, version = 0x0004, abbr_offset = 0x0000, addr_size = 0x04
+
+; ModuleID = '/tmp/a.cpp'
+source_filename = "/tmp/a.cpp"
+target datalayout = "e-m:o-p:32:32-Fi8-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32"
+target triple = "armv4t-darwin-watchos2.0.0"
```
Comment (adrian-prantl) [inline llvm/test/CAS/watchos-cas.ll]: That triple isn't a thing.
`armv7k-apple-watchos2.0.0`

File: llvm/test/CAS/watchos-cas-dwarf5.ll [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+!8 = !{!"clang version 19.0.0git (git@github.com:apple/llvm-project.git b1f1deb2dffc7da3826ef7641006a85b2b65a3e1)"}
+!9 = distinct !DISubprogram(name: "foo", linkageName: "_Z3fooi", scope: !10, file: !10, line: 1, type: !11, scopeLine: 1, flags: DIFlagPrototyped, spFlags: DISPFlagDefinition, unit: !0, retainedNodes: !14)
+!10 = !DIFile(filename: "/tmp/a.cpp", directory: "", checksumkind: CSK_MD5, checksum: "80e82b6b66c04b9cfb438eb7c6672107")
+!11 = !DISubroutineType(types: !12)
+!12 = !{!13, !13}
+!13 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
+!14 = !{}
+!15 = !DILocalVariable(name: "x", arg: 1, scope: !9, file: !10, line: 1, type: !13)
+!16 = !DILocation(line: 1, column: 13, scope: !9)
+!17 = !DILocation(line: 2, column: 12, scope: !9)
```
Comment (adrian-prantl) [inline llvm/test/CAS/watchos-cas-dwarf5.ll]: Can you strip out all unnecessary debug info metadata?

**Comment by rastogishubham** [comment]
@swift-ci please test llvm

**Comment by rastogishubham** [comment]
@swift-ci please test llvm

**Review by adrian-prantl — APPROVED** [review]

**Comment by cachemeifyoucan** [comment]
> When you test 32 bit arches, can you pick something more relevant than armv4t? Also it is always good to test arm64_32 since it is such an outliner and might have edge cases.

I was expecting some tests in regular `armv7` targets while add additional tests for arm64_32. Otherwise LGTM

**Review by felipepiovezan — APPROVED** [review]
LGTM!
